### DISCUSSION TITLE
Open file without losing sidebar focus

### DISCUSSION THREAD

**Discussion by krmbzds** [nvim-neo-tree/neo-tree.nvim#249] [Q&A] [answered]
I want to open a bunch of files from the sidebar without losing focus.

Something like this:
```lua
mappings = {
  ["<Tab>"] = "open_nofocus",
}
```
I read the docs but could not find something equivalent.

## Use case:

Opening 3 adjacent files from sidebar.

### Expected Outcome

I'm hovering over a file and I want to open it along with the 2 files below it in the sidebar.

<kbd>tab</kbd><kbd>j</kbd>
<kbd>tab</kbd><kbd>j</kbd>
<kbd>enter</kbd>

I opened 2 files without losing focus and jumped to the third one.

### Current State

<kbd>enter</kbd><kbd>leader</kbd><kbd>0</kbd>
<kbd>j</kbd><kbd>enter</kbd><kbd>leader</kbd><kbd>0</kbd>
<kbd>j</kbd><kbd>enter</kbd><kbd>leader</kbd><kbd>0</kbd>

Note: <kbd>leader</kbd><kbd>0</kbd> is set to `:Neotree focus`

**Comment by cseickel** [comment, marked as answer]
The easiest way to do that is to create a custom mapping to your own function that calls a built-in open command, then toggles back to the tree:

```lua
  require("neo-tree").setup({
    window = {
      mappings = {
        ["O"] = function (state)
          state.commands["open"](state)
          vim.cmd("Neotree reveal")
        end,

...
```

**Reply by krmbzds** [reply]
Works beautifully! Thanks!

**Reply by gegoune** [reply]
Could you maybe expose lua API so things like `vim.cmd("Neotree reveal")` could be replaced with function call?

**Reply by gegoune** [reply]
Also, noticed that it causes my cursor to appear in the tree for a split second. If you don't remember, I am using events (enter, leave) to change cursor settings to hide it so I only get cursrorline displayed.

nvim-tree does something different, I think, and with similar set up (just using straight up autocmds) there is no brief appearance of cursor (relevant code is here I think [actions/open-file.lua#L152](https://github.com/kyazdani42/nvim-tree.lua/blob/master/lua/nvim-tree/actions/open-file.lua#L152)). Can you think of a way to work around it?

**Reply by cseickel** [reply]
For the lua api, yes that exists. The lua version of the command is:

```lua
    require("neo-tree.command").execute(args)
```

where `args` is a table that can have any of the key=value options that the command takes. For example:
```lua
    require("neo-tree.command").execute({
        action = "focus",
        source = "filesystem",
        -- etc
    })
```

But in reality, since all you want to do is switch back to the window, the most efficient way of doing that is with:
```lua
    vim.api.nvim_set_current_win(state.winid)
```

**Reply by gegoune** [reply]
I like the vim.api call best, thanks for that, neat idea. I just noticed that if I open file using that method cursor blinks while I switch between windows. But once buffer exists switching to that does not blink the cursor.

**Reply by cseickel** [reply]
> Also, noticed that it causes my cursor to appear in the tree for a split second. If you don't remember, I am using events (enter, leave) to change cursor settings to hide it so I only get cursrorline displayed.

If you want to get fancy with it, you can just handle the `"open_file_requested"` event instead. That way you can open the file without having to actually switch windows at all. The downside there is that you have to implement your own window selection logic, because the built-in logic will fall back to just cycling through open windows looking for one that isn't neo-tree.

https://github.com/nvim-neo-tree/neo-tree.nvim/wiki/Recipes#custom-window-chooser-for-file-open-commands

**Reply by nyngwang** [reply]
To @krmbzds, @cseickel:

FYI: The current solution is technically not correct: If the node under the cursor is a directory(folder), then the snippet will close the tree instead of just expanding that folder.

So if the current node is a folder the subsequent `vim.cmd("Neotree reveal")` is not needed.

**update**: I found a solution by referencing an old issue. This should work:

- but I must put a side-node here: calling `Neotree ...` twice will cause a "blink", and I don't like it. But it's acceptable since the functionality is really good.

```lua
['YOUR_SHORTCUT_HERE'] = function (state)
  local node = state.tree:get_node()
  if require("neo-tree.utils").is_expandable(node) then
    state.commands["toggle_node"](state)
  else
    state.commands['open'](state)
    vim.cmd('Neotree reveal')                  
    -- vim.cmd('Neotree float toggle reveal_force_cwd dir=' .. vim.fn.getcwd()) -- I use this one.
  end
end,
```

**Reply by krmbzds** [reply]
@nyngwang I mapped both solutions to different keys and they behaved identically (visually at least). BTW, I had no blinks with your solution (`NVIM v0.7.0-dev+1395-ge135adcb8`). I agree `reveal` isn't necessary if current node is a folder. I added your solution to [Tips](https://github.com/nvim-neo-tree/neo-tree.nvim/wiki/Tips/) page as well. I am currently trying this one out. Thanks!

**Comment by krmbzds** [comment]
This is such a common use-case it might be default behavior IMHO. I switched from `nvim-tree` and got used to [this behavior](https://github.com/kyazdani42/nvim-tree.lua/blob/master/lua/nvim-tree/actions/init.lua#L22) as I was using that. Just my 2c.

Another such feature is [Navigation with HJKL](https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/163#discussioncomment-2274052). I was so happy when I found this one because it was how I expected the tree to behave.

With these two, `neo-tree` is feature-complete for me ATM. I am no longer having any of the issues I had with `nvim-tree` and `neo-tree` works beautifully.

Thank you for the great plugin @cseickel! :heart:

**Comment by krmbzds** [comment]
Updated [Tips](https://github.com/nvim-neo-tree/neo-tree.nvim/wiki/Tips/) page so it's easier to find these.

**Reply by nyngwang** [reply]
Really a good question, thanks for asking!

**Comment by HaloGamer33** [comment]
I got an error everytime that I tried to use the tab key to open the files, for some reason it worked after I removed the `reveal` from `Neotree reveal` I write this if somebody runs into this problem in the future.

This is the code that works for me, like I said, only removed one word but if someone wants to copy and paste it, here it is.
```lua
["<tab>"] = function(state)
  state.commands["open"](state)
  vim.cmd("Neotree")
end,
```

Sidenote: The error only occurs when using `<tab>` as the key to activate the function, it works fine with some other keys, letters seem to work everytime.

**Reply by krmbzds** [reply]
Thank you for sharing this tip!
